test/test_device_spmv.cu

```diff
@@ -0,0 +1,589 @@
+// Compare the reference and cub output vectors.
+// Use fuzzy check for floating point values.
+template <typename ValueT>
+bool compare_results(std::true_type /* is_fp */,
+                     const thrust::host_vector<ValueT>& h_vec1,
+                     const thrust::device_vector<ValueT>& d_vec2)
+{
+  thrust::device_vector<ValueT> d_vec1(h_vec1);
+  auto err = thrust::mismatch(d_vec1.cbegin(),
+                              d_vec1.cend(),
+                              d_vec2.cbegin(),
+                              fp_almost_equal_functor<ValueT>{});
+  if (err.first == d_vec1.cend() || err.second == d_vec2.cend())
+  {
+    return true;
+  }
+  else
+  {
+    thrust::host_vector<ValueT> h_vec2(d_vec2);
+    const auto idx = thrust::distance(d_vec1.cbegin(), err.first);
+    std::cerr << "Mismatch at position " << idx << ": "
+              << print_cast(ValueT{h_vec1[idx]}) << " vs "
+              << print_cast(ValueT{h_vec2[idx]}) << std::endl;
+    return false;
+  }
+};
+
+template <typename ValueT>
+bool compare_results(std::false_type /* is_fp */,
+                     const thrust::host_vector<ValueT>& h_vec1,
+                     const thrust::device_vector<ValueT>& d_vec2)
+{
+
+  thrust::device_vector<ValueT> d_vec1(h_vec1);
+  auto err = thrust::mismatch(d_vec1.cbegin(), d_vec1.cend(), d_vec2.cbegin());
+  if (err.first == d_vec1.cend() || err.second == d_vec2.cend())
+  {
+    return true;
+  }
+  else
+  {
+    thrust::host_vector<ValueT> h_vec2(d_vec2);
+    const auto idx = thrust::distance(d_vec1.cbegin(), err.first);
+    std::cerr << "Mismatch at position " << idx << ": "
+              << print_cast(ValueT{h_vec1[idx]}) << " vs "
+              << print_cast(ValueT{h_vec2[idx]}) << std::endl;
+    return false;
+  }
+}
+
+//==============================================================================
+// Generate a random host_csr_matrix<ValueT> with the specified dimensions.
+// target_fill_ratio is the target fraction of non-zero elements (may be more
+// or less in the output).
+template <typename ValueT>
+host_csr_matrix<ValueT> make_random_csr_matrix(int num_rows,
+                                               int num_cols,
+                                               float target_fill_ratio)
+{
+  host_csr_matrix<ValueT> mat{num_rows, num_cols};
+
+  for (int row = 0; row < num_rows; ++row)
+  {
+    for (int col = 0; col < num_cols; ++col)
+    {
+      const bool is_non_zero = RandomValue<float>(1.f) < target_fill_ratio;
+      if (!is_non_zero)
+      {
+        continue;
+      }
+
+      ValueT value{};
+      do
+      {
+        if (std::is_floating_point<ValueT>::value)
+        { // Keep fp numbers somewhat small; otherwise we run into issues with
+          // nans/infs
+          value = RandomValue(static_cast<ValueT>(200)) -
+                  static_cast<ValueT>(100);
+        }
+        else
+        {
+          InitValue(RANDOM, value);
+        }
+      } while (value == ValueT{});
```

Comment:
Gotcha. I'm not at all familiar with this algorithm or it's implications, and wasn't sure whether having "zero non-zeros"  in the data would cause issues. If this is a common case, I agree that we should test it.